src/v/pandaproxy/schema_registry/service.cc

```diff
@@ -109,6 +112,32 @@ ss::future<> service::create_internal_topic() {
     // TODO(Ben): Validate the _schemas topic
 }
 
+ss::future<> service::fetch_internal_topic() {
+    constexpr static auto offset_req = []() {
+        return kafka::list_offsets_request{
+          .data = {.topics{
+            {{.name{schemas_tp.topic},
+              .partitions{
+                {{.partition_index{schemas_tp.partition},
+                  .max_num_offsets = 128}}}}}}}};
+    };
+    auto offset_res = co_await _client.local().dispatch(offset_req);
+    auto max_offset = offset_res.data.topics[0].partitions[0].offset;
+    vlog(plog.debug, "Schema registry: _schemas max_offset: {}", max_offset);
+
+    client_fetcher reader{
+      _client.local(), schemas_tp, model::offset{0}, max_offset};
+    consume_to_store consumer{_store};
+
+    co_await ss::do_until(
+      [this, &reader]() {
+          return _gate.is_closed() || reader.is_end_of_stream();
+      },
+      [&consumer, &reader]() {
+          return std::move(reader).consume(consumer, model::no_timeout);
```

Comment:
something about this looks a little off. the body of the do_until does std::move(reader) but could be repeated, so appears like a use-after-free (even if std::move(reader)) doesn't clobber stuff.

generally I think that if client_fetcher is designed properly you don't need the do_until loop. the looping should be handled by consume internally.

Reply:
Fixed, thanks.